rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,275 @@
+that add specific fulltext search indexes to an existing subgraph. 
+
+Example fulltext search definition:
+
+```graphql
+type _Schema_ 
+  @fulltext(
+    name: "media"
+    ...
+  )
+  @fulltext(
+    name: "search",
+    language: "english",
+    algorithm: "ranked",
+    include: [
+      {
+        entity: "Band",
+        fields: [
+          { name: "name", weight: 5 },
+        ]
+      },
+      {
+        entity: "Album",
+        fields: [
+          { name: "title", weight: 5 },
+        ]
+      },
+      {
+        entity: "Musician",
+        fields: [
+          { name: "name", weight: 10 },
+          { name: "bio", weight: 5 },
+        ]
+      }
+    ]
+  )
+```
+
+The schema generated from the above definition: 
+```graphql
+union FulltextMediaResultItem = ...
+union FulltextSearchResultItem = Band | Album | Musician
+type Query {
+  media...
+  search(text: String!, first: Int, skip: Int, block: Int): [FulltextSearchResultItem!]!
+}
+```
+
+### GraphQL Query interface
+
+End users of the subgraph will have access to the queries generated for the
+fulltext search indexes alongside the other queries generated for each 
+entity in the subgraph.  In the case of a fulltext search index that spans
+multiple entities, [inline fragments](https://graphql.org/learn/queries/#inline-fragments) 
+may be used in the query to deal with the polymorphic result items. In the 
+front-end the `__typename` in each result item can be used to render the 
+components accordingly.  
+
+In the "text" parameter supplied to the query there will be one special 
+operator supported, the proximity operator. Several operators are available: 
+and, or, and proximity (`&`, `|`, `<->`.) The proximity operator allows 
+one to specify max distance between search terms: 3 words apart → `<3>`. 
+
+Example query using inline fragments and the proximity operator: 
+```graphql
+query {
+  search(text: "Bob<3>run") {
+    __typename
+    ... on Band { name label { id } }
+    ... on Album { title numberOfTracks }
+    ... on Musician { name bio }
+  }
+}
+```
+
+### Tools and Design
+
+Fulltext search query system implementations often involve specific systems 
+for storing and querying the text documents; however, in an effort to reduce 
+system complexity and feature implementation time I propose starting with the 
+fulltext search features built in to PostgreSQL.
+
+A FullText search field will get its own column in the Entity table just 
+like any other field; however, the data stored will be the result of the 
+lexical, morphological analysis of text documents performed on the fields 
+included in the index. The fulltext search field will be created using the 
+Postgres ts_vector function and will be indexed using a GIN index. The subgraph 
+developer will define a ranking algorithm to be used to sort query results,
+so the end-user facing API remains easy to use without any requirement to 
+understand the ranking algorithms.   
```

Comment:
The last time we discussed this, we discussed using a separate table for this data. I'm ok either way, one extra column per entity per fulltext search field could work as well. Just wanted to double-check which one we're choosing.